# List Comprehensions Lab

Complete the following set of exercises to solidify your knowledge of list comprehensions.

In [4]:
import os
import numpy as np
import pandas as pd

### 1. Use a list comprehension to create and print a list of consecutive integers starting with 1 and ending with 50.

In [ ]:
print([i+1 for i in range(50)])

In [20]:
x = [i for i in range(1, 51, 1)]
print(x)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]


### 2. Use a list comprehension to create and print a list of even numbers starting with 2 and ending with 200.

In [27]:
print([i+2 for i in range (200) if i % 2 == 0])

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200]


### 3. Use a list comprehension to create and print a list containing all elements of the 10 x 4 Numpy array below.

In [21]:
a = np.array([[0.84062117, 0.48006452, 0.7876326 , 0.77109654],
       [0.44409793, 0.09014516, 0.81835917, 0.87645456],
       [0.7066597 , 0.09610873, 0.41247947, 0.57433389],
       [0.29960807, 0.42315023, 0.34452557, 0.4751035 ],
       [0.17003563, 0.46843998, 0.92796258, 0.69814654],
       [0.41290051, 0.19561071, 0.16284783, 0.97016248],
       [0.71725408, 0.87702738, 0.31244595, 0.76615487],
       [0.20754036, 0.57871812, 0.07214068, 0.40356048],
       [0.12149553, 0.53222417, 0.9976855 , 0.12536346],
       [0.80930099, 0.50962849, 0.94555126, 0.33364763]])

In [48]:
flatten = [value for sublist in a for value in sublist]
print(flatten)

[0.84062117000000003, 0.48006451999999999, 0.78763260000000002, 0.77109654000000005, 0.44409792999999997, 0.090145160000000002, 0.81835917000000002, 0.87645455999999999, 0.7066597, 0.096108730000000003, 0.41247947000000001, 0.57433389000000001, 0.29960807, 0.42315023000000002, 0.34452557, 0.47510350000000001, 0.17003562999999999, 0.46843997999999998, 0.92796257999999998, 0.69814653999999998, 0.41290051, 0.19561070999999999, 0.16284783, 0.97016248000000005, 0.71725408000000002, 0.87702738000000002, 0.31244594999999997, 0.76615487000000004, 0.20754036000000001, 0.57871812, 0.072140679999999999, 0.40356048, 0.12149553, 0.53222417, 0.9976855, 0.12536346000000001, 0.80930099, 0.50962848999999999, 0.94555126, 0.33364762999999997]


### 4. Add a condition to the list comprehension above so that only values greater than or equal to 0.5 are printed.

In [39]:
flatten = [value for sublist in a for value in sublist if value >= 0.5]
print(flatten)

[0.84062117000000003, 0.78763260000000002, 0.77109654000000005, 0.81835917000000002, 0.87645455999999999, 0.7066597, 0.57433389000000001, 0.92796257999999998, 0.69814653999999998, 0.97016248000000005, 0.71725408000000002, 0.87702738000000002, 0.76615487000000004, 0.57871812, 0.53222417, 0.9976855, 0.80930099, 0.50962848999999999, 0.94555126]


### 5. Use a list comprehension to create and print a list containing all elements of the 5 x 2 x 3 Numpy array below.

In [41]:
b = np.array([[[0.55867166, 0.06210792, 0.08147297],
        [0.82579068, 0.91512478, 0.06833034]],

       [[0.05440634, 0.65857693, 0.30296619],
        [0.06769833, 0.96031863, 0.51293743]],

       [[0.09143215, 0.71893382, 0.45850679],
        [0.58256464, 0.59005654, 0.56266457]],

       [[0.71600294, 0.87392666, 0.11434044],
        [0.8694668 , 0.65669313, 0.10708681]],

       [[0.07529684, 0.46470767, 0.47984544],
        [0.65368638, 0.14901286, 0.23760688]]])

In [61]:
flatten2 = [value for sublist in b for value in sublist]
flatten3 = [value for sublist in flatten2 for value in sublist]
print(flatten3)

[0.55867166000000001, 0.062107919999999997, 0.081472970000000006, 0.82579068, 0.91512477999999997, 0.068330340000000003, 0.054406339999999997, 0.65857692999999995, 0.30296619000000002, 0.067698330000000001, 0.96031862999999995, 0.51293743000000003, 0.091432150000000004, 0.71893381999999995, 0.45850679, 0.58256463999999997, 0.59005653999999996, 0.56266457000000003, 0.71600293999999998, 0.87392665999999997, 0.11434044, 0.86946679999999998, 0.65669312999999996, 0.10708681, 0.075296840000000004, 0.46470767000000002, 0.47984544000000001, 0.65368638000000001, 0.14901286, 0.23760687999999999]


### 5. Add a condition to the list comprehension above so that the last value in each subarray is printed, but only if it is less than or equal to 0.5.

In [57]:
flatten2 = [value[-1] for sublist in b for value in sublist if value[-1] >= 0.5]
print(flatten2)

[0.51293743000000003, 0.56266457000000003]


### 6. Use a list comprehension to select and print the names of all CSV files in the */data* directory.

In [5]:
path = "../data"

def listfileNames(path):
    return [filename for filename in os.listdir(path) if filename[-4:] == ".csv"]

print(listfileNames(path))


['sample_file_6.csv', 'sample_file_2.csv', 'sample_file_0.csv', 'sample_file_8.csv', 'sample_file_5.csv', 'sample_file_3.csv', 'sample_file_7.csv', 'sample_file_9.csv', 'sample_file_4.csv', 'sample_file_1.csv']


### 7. Use a list comprehension and the Pandas `read_csv` and `concat` methods to read all CSV files in the */data* directory and combine them into a single data frame. Display the top 10 rows of the resulting data frame.

In [9]:
#pd.read_csv("../data")

data_sets = [pd.read_csv(os.path.join('../data', f)) for f in listfileNames]
data = pd.concat(data_sets, axis=0)
data.head(10)


TypeError: 'function' object is not iterable

### 8. Use a list comprehension to select and print the column numbers for columns from the data set whose median is less than 0.48.

### 9. Use a list comprehension to add a new column (20) to the data frame whose values are the values in column 19 minus 0.1. Display the top 10 rows of the resulting data frame.

### 10. Use a list comprehension to extract and print all values from the data set that are between 0.7 and 0.75.